app/models/topic_converter.rb

```diff
@@ -75,7 +75,6 @@ def update_user_stats
   def add_allowed_users
     @topic.posts.where(deleted_at: nil).each do |p|
       user = User.find(p.user_id)
-      @topic.topic_allowed_users.build(user_id: user.id) unless @topic.topic_allowed_users.where(user_id: user.id).exists?
```

Comment:
This is adding `user.id` not `@user.id` this is the user associated with the post. 

So instead, keep track of all the new users in a hash or array and do the addition the end.

```
users = {}
if ...  
     # allows for deleted users
     user = users[p.user_id] ||= User.find_by(id: p.user_id)
     if user
       user.user_stat...
     end
end

finally ensuring al these new users exist in topic allowed users...

Reply:
You are right, my bad. What do you think of instead of building a hash of users, remove duplications just before creation?